torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,489 @@
+
+import torch
+import torch.distributed as dist
+from torch.nn import Parameter
+from torch._six import container_abcs
+from torch.optim import Optimizer
+
+__all__ = ["ZeroRedundancyOptimizer"]
+
+
+if TYPE_CHECKING:  # pragma: no cover
+    from torch.optim.optimizer import _params_t
+else:
+    _params_t = Any
+
+
+# Credits:  classy_vision/generic/distributed_util.py
+def _recursive_copy_to_device(value: Any, non_blocking: bool, device: torch.device) -> Any:
+    """
+    Recursively searches lists, tuples, dicts and copies tensors to device if
+    possible. Non-tensor values are passed as-is in the result.
+
+    .. note:  These are all copies, so if there are two objects that reference
+    the same object, then after this call, there will be two different objects
+    referenced on the device.
+    """
+
+    if isinstance(value, torch.Tensor):
+        return value.to(device, non_blocking=non_blocking)
+
+    if isinstance(value, (list, tuple)):
+        values = [_recursive_copy_to_device(val, non_blocking=non_blocking, device=device) for val in value]
+        return values if isinstance(value, list) else tuple(values)
+
+    if isinstance(value, container_abcs.Mapping):
+        return {
+            key: _recursive_copy_to_device(val, non_blocking=non_blocking, device=device) for key, val in value.items()
+        }
+
+    return value
+
+
+def _get_global_rank(group: Any, rank: int) -> int:
+    return rank if group is dist.group.WORLD else dist.distributed_c10d._get_global_rank(group, rank)  # type: ignore
+
+
+class ZeroRedundancyOptimizer(Optimizer):
+    """Wraps an arbitrary :class:`optim.Optimizer <torch.optim.Optimizer>`
+    optimizer and shards its state as described by ZeRO_.
+    ::
+
+        opt = ZeroRedundancyOptimizer(params, optim=torch.optim.Adam, lr=0.01)
+
+    .. _ZeRO: https://arxiv.org/abs/1910.02054
+
+    We use a greedy algorithm to pack a number of parameters at each rank.
+    Each parameter belongs to a single rank and is not divided among ranks.
+    The partition is arbitrary and does not correspond to the information flow for instance.
+
+    After each rank completed their parameter update, they broadcast
+    the new version of the parameters to all other ranks to synchronize
+    the parameters for next round forward/backward computation.
+
+    Arguments:
+        params (list of tensors):
+            parameters to be optimized
+    Keyword Args:
+        optim (torch.nn.Optimizer):
+            optimizer to shard
+        group (group):
+            torch.distributed group (default: group.WORLD)
+        bucket_cap (int):
+            the size of the buffer used to batch the small parameter tensors, in number of elements (default 16M).
+    """
+
+    def __init__(
+        self,
+        params: _params_t,
+        optim: Type[Optimizer],
+        group: Optional[Any] = None,
+        bucket_cap_kb: int = 2 ** 24,
+        **default: Any,
```

Comment:
do we want to mention in the doc that the trailing args will be forwarded to the optimizer ctor?

Reply:
ah yes, good point ! I can fix that